README.rst
nxt-graph
==========================================

Introduction
------------

Generic javascript graph library.

Prerequisities
--------------

You must use AngularJS in order to use nxt-graph (get it from elsewhere). Also include d3, nv, nxt-graph js and css (embedded in nxt-graph) in your document::


  <script type="text/javascript" src="{{ STATIC_URL }}angular/angular.min.js"></script>
  <script type="text/javascript" src="{{ STATIC_URL }}vendor/d3js/d3.v3.min.js?{{ random_string }}"></script>
  <script type="text/javascript" src="{{ STATIC_URL }}vendor/nvd3/nv.d3.js?{{ random_string }}"></script>
  <script type="text/javascript" src="{{ STATIC_URL }}nxt_graph/nxt-graph.js?{{ random_string }}"></script>
  <link href="{{ STATIC_URL }}vendor/nvd3/source/nv.d3.css" rel="stylesheet"/>




Usage
-----

Make directives and modify the infourl from Angular. The nxt-timeseries and nxt-profile will automagically become fancy nv d3 graphs! Even more: every time when the inner url is updated, the graph will refresh itself. This way you can get realtime moving graphs!


Example::


  app.controller("InfoPoint", ["$scope", function($scope) {
      $scope.$on("infopoint", function(message, content) {
          $scope.$apply(function() {
              console.log("open box infopoint");
              var lonlat = content.point;
              var $layer = document.getElementsByClassName("workspace-wms-layer")[0];  // there is only one
              $scope.infourl = $layer.dataset['workspaceWmsUrl'].split('/wms')[0] + '/data?' + "REQUEST=gettimeseries&LAYERS=" + content.loaded_model + "&SRS=EPSG:4326&POINT="+lonlat.lng.toString() + ',' + lonlat.lat.toString();
          });        
      });

  }]);      


Timeseries
==========

Example::

  <nxt-timeseries url="{[{ infourl }]}"></nxt-timeseries>

The url must contain data in the format::

  {"timeseries": [["2012-01-01T00:01:00", -0.946], ["2012-01-01T00:02:00", -0.946], 
                  ["2012-01-01T00:03:00", -0.946], ["2012-01-01T00:04:00", -0.946], 
                  ["2012-01-01T00:05:00", -0.946], ["2012-01-01T00:06:00", -0.946]]}


Profiles
========

Example::

  <nxt-profile url="{[{ infourl }]}"></nxt-profile>

The url must contain data in the format::

  {"depth": [[0.0, 0.2673], [1e-05, 0.2673], [3e-05, 0.2673], [3e-05, 0.2673], [4e-05, 0.2773]], 
   "waterheight": [[0.0, 0.2673], [1e-05, 0.2673], [3e-05, 0.2673], [3e-05, 0.2673], [4e-05, 0.2773]], 
   "bathymetry": [[0.0, 0.2673], [1e-05, 0.2673], [3e-05, 0.2673], [3e-05, 0.2673], [4e-05, 0.2773]]}


CHANGES.rst
Changelog of nxt-graph
===================================================


0.16 (unreleased)
-----------------

- Nothing changed yet.


0.15 (2015-01-14)
-----------------

- Update bootstrap.py.

- Adjusted the time series so that small changes are not magnified.


0.14 (2014-11-27)
-----------------

- Remove objects from dom to prevent memory leaks, as well as not adding circles
 (patched included nvd3).

- Updated crosssection as well as timeseries graphs.


0.12 (2014-10-07)
-----------------

- Fixed default profile in case of server errors.


0.11 (2014-07-14)
-----------------

- Better y scaling for profiles.


0.10 (2014-03-31)
-----------------

- Changed tooltip for profile.

- Removed carsten timeseries.


0.9 (2014-03-13)
----------------

- Changed tooltip display.


0.8 (2014-03-12)
----------------

- Changed input for nxt crosssection: added ground water and renamed fields.


0.7 (2014-02-17)
----------------

- Added requestAnimationFrame. There is still a problem when requesting graphs repeatedly.


0.6 (2014-02-06)
----------------

- Bugfix and re-enabled busy=true.


0.5 (2014-02-05)
----------------

- Disabled 'busy = true': multiple graphs in one popup are now possible, because of this.


0.4 (2014-01-13)
----------------

- Enabled 'readyForNext': graphs are now always updated to the newest available version.


0.3 (2013-09-19)
----------------

- Changed axis label Time (HH:MM) -> Time.

- Changed water/land to depth/elevation.


0.2 (2013-08-20)
----------------

- Improvements in axis and layout.

- Added nxtCarstenTimeSeries.

- Slightly changed nv.d3.css to prevent slash with leaflet.


0.1 (2013-07-29)
----------------

- You can look at time series and cross sections, they will update every time when the url is updated.

- Initial project structure created with nensskel 1.32.


nxt_graph/static/nxt_graph/nxt-graph.js
//nxt-graph.js

// create the directives as re-usable components
app
    .directive('nxtTimeseries', function($http) {
        return {
            restrict: 'E',
            replace: true,
            scope: {
                'url': '@',
                'title': '@',
                'unit': '@'
            },
            template: '<svg></svg>',
            link: function(scope, element, attrs) {
                var readyForNext = null;
                var busy = false;
                var getData = function(url, fn){
                    //console.log(url);
                    $.ajax({
                            url: url,
                            type: 'GET',
                            dataType: 'json',
                            success: function(data) {
                                //console.log('data!!!', data);
                                var formatted = [{
                                            "key": "timeseries",
                                            "values": data['timeseries']
                                        }];
                                fn(formatted);

                                console.log('readyForNext: ', readyForNext);
                                if (readyForNext !== null) {
                                    setTimeout(function() {
                                        requestAnimationFrame(function() {
                                            // possibly a user does not see the very
                                            // latest graph...
                                            if (readyForNext !== null) {
                                                console.log("ReadyForNext!!");
                                                var next = readyForNext;
                                                readyForNext = null;
                                                getData(next, addGraph);
                                            }
                                            busy = false;
                                        });
                                    }, 1000);  // wait a while before accepting new
                                } else {
                                    busy = false;
                                }
                            },
                            error: function (data) {
//                                console.log('error!!!', data);
                                var empty = [{"key": "timeseries",
                                            "values": [['1979-05-25 14:30', 0], ['1979-05-25 14:31', 0]]}];
                                fn(empty);
                                setTimeout(function() {
                                    requestAnimationFrame(function() {
                                        busy = false;
                                    });
                                }, 1000);  // wait a while before accepting new
                            }
                    });  // $.ajax
                }
                var addGraph = function(formatted) {
                    var result = nv.addGraph(function() {
                        var chart = nv.models.lineChart()
                                      .x(function(d) { return Date.parse(d[0]) })
                                      .y(function(d) { return d[1] })
                                      .clipEdge(true)
                                      .tooltipContent(function(key, y, e, graph) {
                                        var header = (scope.title != undefined) ? scope.title : key;
                                        if ((scope.unit != undefined) && (scope.unit != '') && (scope.unit != null) ) {
                                            header = header + ' - ' + scope.unit;
                                        }
                                        return '<h3 class="graph-header">' + header + ' </h3>' +
                                               '<p>' +  e + '</p>'
                                    });
                        var epoch = 0;
                        try {
                            // try to get the startdate.
                            epoch = +Date.parse(formatted[0].values[0][0]);
                        } catch(err) {
                        }
//                        console.log('epoch for this graph is ', epoch);
                        chart.xAxis
                            .axisLabel('Time')
                            .tickFormat(function(d) {
                                //var hours = +(d- new Date("2012-01-01")) / 1000 / 60 / 60;
                                //console.log('debug ', ((+d) - epoch));
                                var minutes = ((+d) - epoch)  / 1000 / 60;
                             //return Math.round(hours*10)/10;
                                var minutes_mod = Math.floor(minutes%60);
                                if (minutes_mod < 10) {minutes_mod = '0' + minutes_mod;}
                             return Math.floor(minutes/60)+':'+minutes_mod;
                             //return d3.time.format('%d %H:%M')(new Date(d))
                           });

                        chart.yAxis
                             .axisLabel('Depth (m)')
                             .tickFormat(d3.format(',.2f'));

                        // Adjust y domain so that small changes are not magnified.
                        var yMin = d3.min(formatted[0].values, function(d) { return d[1]; });
                        var yMax = d3.max(formatted[0].values, function(d) { return d[1]; });
                        var yDelta = yMax - yMin;
                        if (yDelta < 0.01) {
                            console.log("Adjust y axis, because yDelta =", yDelta, "< 0.01");
                            chart.yDomain([yMin-yDelta*5, yMax+yDelta*5]);
                        } else {
                            console.log("Adjust y axis, with yDelta =", yDelta);
                            chart.yDomain([yMin-yDelta*0.05/yDelta, yMax+yDelta*0.05/yDelta]);
                        }
                        chart.showLegend(false);

                        //console.log('element', $(element).attr('id'), element);
                        // Make sure your context as an id or so...
                        d3.select(element.context)
                          .datum(formatted)
                            .transition().duration(500).call(chart);

                        nv.utils.windowResize(chart.update);
                        scope.current_chart = chart;
                        return chart;

                    });  // nv.addGraph
                    return result;
                };

                var set_null_on_selection = function(selection) {
                    for (o in selection.remove()) {
                        o = null;
                    };
                }

                scope.$watch('url', function (url) {
                    if ((url !== '') ) {
                        if (busy) {
                            // We don't have time for it now, but later you want
                            // the latest available graph.
                            readyForNext = url;
                            return;
                        }
                        if (scope.current_chart !== null) {
                            // uncomment this to use slightly less memory, but the line
                            // graph will flicker on update
/*                            set_null_on_selection(d3.selectAll('svg g.nv-pointpaths path'));
                            set_null_on_selection(d3.selectAll('svg g.nv-pointpaths'));
                            set_null_on_selection(d3.selectAll('svg g.nv-group'));
                            set_null_on_selection(d3.selectAll('svg g.nv-groups'));
                            set_null_on_selection(d3.selectAll('svg g.nv-wrap'));
                            set_null_on_selection(d3.selectAll('svg g.nv-scatterWrap'));
                            set_null_on_selection(d3.selectAll('svg g.nv-linesWrap'));
                            set_null_on_selection(d3.selectAll('svg g.nv-axis'));
                            set_null_on_selection(d3.selectAll('svg g.nvd3'));
*/
                            scope.current_chart = null;

                            window.nv.charts = {};
                            window.nv.graphs = [];
                            window.nv.logs = {};
                            window.onresize = null;
                        }

                        // NB: busy = true commented because it doesn't allow
                        // for multiple graphs in the same popup. This is needed
                        // for example for the graphs in the threedi orifice popup.
                        busy = true;

                        //console.log('busy', busy);
                        getData(url, addGraph);

                    }
                });  // scope.watch
            }
        }
    });


app
    .directive('nxtCrossSection', function($http) {
        var busy = false;
        return {
            restrict: 'E',
            replace: true,
            scope: {
                'url': '@',
                'unit': '@'
            },
            template: '<svg></svg>',
            link: function(scope, element, attrs) {
                scope.current_chart = null;
                scope.to_be_deleted = null;
                var getData = function(url, fn){
                    $.ajax({
                            url: url,
                            success: function(data) {
                                var formatted = [{
                                  "key": "offset",
                                  "values": data.offset,
                                  "color": "#ffffff"
                                },{
                                  "key": "groundwater_delta",
                                  "values": data.groundwater_delta,
                                  "color": "#2D8265"
                                },{
                                  "key": "elevation",
                                  "values": data.bathymetry_delta,
                                  "color": "#2C9331"
                                },{
                                  "key": "depth",
                                  "values": data.depth,
                                  "color": "LightSkyBlue"
                                }];
                                //console.log('formatte 1', formatted, data);
                                // new graph and contents
                                fn(formatted, data.summary);

                                setTimeout(function() {
                                    requestAnimationFrame(function() {
                                        busy = false;
                                    });
                                }, 600);
                            },
                            error: function (data) {
                                var empty = [{
                                    "key": "offset",
                                    "values": [[0, 0], [1/111, 0]],
                                    "color": "#ffffff",
                                    "opacity": "0.5"
                                },{
                                    "key": "groundwater_delta",
                                    "values": [[0, 0], [1/111, 0]],
                                    "color": "#2D8265",
                                    "opacity": "0.5"
                                },{
                                    "key": "elevation",
                                    "values": [[0, 0], [1/111, 0]],
                                    "color": "#2C9331"
                                },{
                                  "key": "depth",
                                  "values": [[0,0], [1/111, 0]],
                                  "color": "LightSkyBlue"
                                }];
                                fn(empty);
                                setTimeout(function() {
                                    requestAnimationFrame(function() {
                                        busy = false;
                                    });
                                }, 600);
                            }
                    });  // $.ajax
                }
                var addGraph = function(formatted, summary) {
                    nv.addGraph(function() {
                        //console.log('scope.url2 ', scope.url, '-', scope_url);
                        //console.log('formatted 2', formatted);


                        //console.log("dataaa", data, formatted);
                        // 2 * pi * r / 360 = 111 km per degrees, approximately
                        var chart = nv.models.stackedAreaChart()
                        .height(310).width(380)
                        //var chart = nv.models.lineChart()
                                      .x(function(d) { return 111*d[0] })
                                      .y(function(d) { return d[1] })
                                      .clipEdge(true)
                                      .tooltipContent(function(key, y, e, graph) {
                                        var header = (scope.title != undefined) ? scope.title : key;
                                        if ((scope.unit != undefined) && (scope.unit != '') && (scope.unit != null) ) {
                                            header = header + ' - ' + scope.unit;
                                        }
                                        var unit_x = 'km';
                                        return '<h3 class="graph-header">' + header + ' </h3>' +
                                               '<p>' + e + ' at ' + y + unit_x + ' </p>';
                                    });

                        chart.xAxis
                            .axisLabel('Distance')
                            .tickFormat(d3.format(',.2f'));

                        chart.yAxis
                            .axisLabel('Depth')
                            .tickFormat(d3.format(',.2f'));
                            // .tickFormat(function(d){return d-5});

                        if (summary !== undefined) {
                            // Get summary data from threedi-wms?
                            var minVal = summary.minimum;
                            var maxVal = summary.maximum;
                            var marginVal = summary.margin;
                        } else {
                            console.log('Please update threedi-wms for a better experience...')
                            // Old threedi-wms...
                            var minVal = d3.min(formatted[0].values, function(d) {return d[1];});

                            //Try stacked.y0????

                            // Maximum heights
                            var sumArray = new Array(formatted[0].values.length);
                            for (var i=0; i<formatted[0].values.length; i++) {
                                sumArray[i] = (minVal +
                                    formatted[1].values[i][1] +
                                    formatted[2].values[i][1] +
                                    formatted[3].values[i][1]);
                            }
                            var maxVal = d3.max(sumArray);
                            var marginVal = 0.1 * (maxVal - minVal);
                        }

                        chart.yDomain([minVal, maxVal + marginVal]);

                        // TODO: disable click handler?
                        // TODO: doesn't rescale correctly when negative 2D is applied??
                        // TODO: BUG: when you click on the graph, it doesn't show
                        // individual profiles correctly anymore
                        // var noChartSelected = function() {
                        //     var cstate = chart.state();
                        //     if (cstate.disabled === undefined) {
                        //         return true;
                        //     }
                        //     if (cstate.disabled.indexOf(true) === -1) {
                        //         return true;
                        //     }
                        //     return false;
                        // };

                        // if (noChartSelected()) {
                        //     chart.yDomain([minVal, maxVal + marginVal]);
                        // }

                        chart.showControls(false);
                        chart.showLegend(false);

                        //console.log('element', $(element).attr('id'), element);
                        // Make sure your context as an id or so...
                        d3.select(element.context)
                          .datum(formatted)
                            .transition().duration(100).call(chart);

                        nv.utils.windowResize(chart.update);
                        scope.current_chart = chart;
                        return chart;

                    });  // nv.addGraph

                };

                var set_null_on_selection = function(selection) {
                    for (o in selection.remove()) {
                        o = null;
                    };
                }

                scope.$watch('url', function (url) {
                    //console.log('profile url update');
                    if (busy) {
                        // Only update if an old request is already finished
                        //console.log("profile: busy!!");
                        return;
                    }
                    if (scope.current_chart !== null) {
                        // Still experimental, these are other options we may have to use.
                        // d3.selectAll('circle').remove();
                        // d3.selectAll('path').remove();
                        // d3.selectAll('svg').remove();
                        // remove stuff so memory clears up a bit.
                        // set_null_on_selection(d3.selectAll('svg g'));
                        // set_null_on_selection(d3.selectAll('svg path'));

                        // remove elements down up to prevent detached elements
                        set_null_on_selection(d3.selectAll('svg g.nv-areaWrap path'));
                        set_null_on_selection(d3.selectAll('svg g.nv-areaWrap'));
                        set_null_on_selection(d3.selectAll('svg g.nv-axis'));
                        set_null_on_selection(d3.selectAll('svg g.nv-stackedWrap'));
                        set_null_on_selection(d3.selectAll('svg g.nv-scatterWrap'));
                        set_null_on_selection(d3.selectAll('svg g.nv-legendWrap'));
                        set_null_on_selection(d3.selectAll('svg g.nv-controlWrap'));
                        set_null_on_selection(d3.selectAll('svg g.nv-wrap'));
                        set_null_on_selection(d3.selectAll('svg g.nvd3'));

                        scope.current_chart = null;

                        window.nv.charts = {};
                        window.nv.graphs = [];
                        window.nv.logs = {};
                        window.onresize = null;
                    }
                    if (url !== '') {
                        //console.log('updating profile graph...');
                        busy = true;

                        getData(url, addGraph);
                    }
                });  // scope.watch
            }
        }
    });
